engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/hydration/batch/NadelBatchHydrationByIndex.kt

```diff
@@ -2,46 +2,77 @@ package graphql.nadel.enginekt.transform.hydration.batch
+import graphql.nadel.enginekt.blueprint.hydration.NadelHydrationActorInputDef
+import graphql.nadel.enginekt.transform.artificial.NadelAliasHelper
 import graphql.nadel.enginekt.transform.hydration.NadelHydrationUtil
-import graphql.nadel.enginekt.transform.hydration.NadelHydrationUtil.getNumberOfInputNodes
 import graphql.nadel.enginekt.transform.hydration.batch.NadelBatchHydrationInputBuilder.getBatchInputDef
 import graphql.nadel.enginekt.transform.result.NadelResultInstruction
 import graphql.nadel.enginekt.transform.result.json.JsonNode
 import graphql.nadel.enginekt.util.AnyList
+import graphql.nadel.enginekt.util.emptyOrSingle
 import graphql.nadel.enginekt.util.isList
 import graphql.nadel.enginekt.util.listOfNulls
-import graphql.nadel.enginekt.util.subListOrNull
 import graphql.nadel.enginekt.util.unwrapNonNull
+import graphql.normalized.ExecutableNormalizedField
 
-internal object NadelBatchHydrationByIndex {
-    fun getHydrateInstructionsMatchingIndex(
-        state: NadelBatchHydrationTransform.State,
-        instruction: NadelBatchHydrationFieldInstruction,
-        parentNodes: List<JsonNode>,
-        batches: List<ServiceExecutionResult>,
-    ): List<NadelResultInstruction> {
+internal class NadelBatchHydrationByIndex private constructor(
+    private val fieldToHydrate: ExecutableNormalizedField,
+    private val aliasHelper: NadelAliasHelper,
+    private val instruction: NadelBatchHydrationFieldInstruction,
+    private val parentNodes: List<JsonNode>,
+    private val batches: List<ServiceExecutionResult>,
+) {
+    companion object {
+        fun getHydrateInstructionsMatchingIndex(
+            state: NadelBatchHydrationTransform.State,
+            instruction: NadelBatchHydrationFieldInstruction,
+            parentNodes: List<JsonNode>,
+            batches: List<ServiceExecutionResult>,
+        ): List<NadelResultInstruction> {
+            return NadelBatchHydrationByIndex(
+                fieldToHydrate = state.hydratedField,
+                aliasHelper = state.aliasHelper,
+                instruction = instruction,
+                parentNodes = parentNodes,
+                batches = batches,
+            ).getHydrateInstructions()
+        }
+    }
+
+    fun getHydrateInstructions(): List<NadelResultInstruction> {
         val isManyInputNodesToParentNodes = isManyInputNodesToParentNodes(instruction)
         val chunker = Chunker(instruction, batches)
 
         return parentNodes
             .map { parentNode ->
+                val inputValues = getInputValues(parentNode)
+
                 NadelResultInstruction.Set(
-                    subjectPath = parentNode.resultPath + state.hydratedField.resultKey,
+                    subjectPath = parentNode.resultPath + fieldToHydrate.resultKey,
                     newValue = if (isManyInputNodesToParentNodes) {
-                        chunker.take(
-                            n = getNumberOfInputNodes(
```

Comment:
So with this change, we still don't ask the underlying service for null nodes. But instead of blindly taking N nodes that the parent node needs, we pass in the input nodes to the `Chunker` which returns extra nulls if it needs to.